arbnode/forwarder.go

```diff
@@ -186,3 +205,162 @@ func (f *TxDropper) StopAndWait() {}
+	if err != nil {
+		return errors.Wrap(err, "unable to create redis coordinator")
+	}
+	return nil
+}
+
+func (f *RedisTxForwarder) retryAfterError() time.Duration {
+	f.errors++
+	retryIn := f.config.RetryInterval * time.Duration(f.errors)
+	if retryIn > f.config.UpdateInterval {
+		retryIn = f.config.UpdateInterval
+	}
+	return retryIn
+}
+
+// returns true when retry interval is saturated and there is a fallback url available
+func (f *RedisTxForwarder) shouldFallbackToStatic() bool {
+	return f.config.RetryInterval*time.Duration(f.errors+1) >= f.config.UpdateInterval &&
+		f.fallbackTarget != "" && f.fallbackTarget != f.currentTarget
+}
+
+func (f *RedisTxForwarder) noError() time.Duration {
+	f.errors = 0
+	return f.config.UpdateInterval
+}
+
+func (f *RedisTxForwarder) getForwarder() *TxForwarder {
+	f.mtx.RLock()
+	defer f.mtx.RUnlock()
+	return f.forwarder
+}
+
+func (f *RedisTxForwarder) setForwarder(forwarder *TxForwarder) {
+	f.mtx.Lock()
+	defer f.mtx.Unlock()
+	f.forwarder = forwarder
+}
+
+// not thread safe vs initialize and itself
+func (f *RedisTxForwarder) update(ctx context.Context) time.Duration {
+	nextUpdateIn := f.noError
+	var newSequencerUrl string
+	var redisErr error
+	if f.redisCoordinator != nil {
+		newSequencerUrl, redisErr = f.redisCoordinator.CurrentChosenSequencer(ctx)
+		if redisErr != nil || newSequencerUrl == "" {
+			if f.shouldFallbackToStatic() && f.fallbackTarget != "" {
+				log.Warn("coordinator failed to find live sequencer, falling back to static url", "err", redisErr, "fallback", f.fallbackTarget)
+				newSequencerUrl = f.fallbackTarget
+				nextUpdateIn = f.retryAfterError
+			} else {
+				log.Warn("coordinator failed to find live sequencer", "err", redisErr)
+				return f.retryAfterError()
+			}
+		}
+	} else {
+		if f.fallbackTarget != "" {
+			log.Warn("redis coordinator not initialized, falling back to static url", "fallback", f.fallbackTarget)
+			newSequencerUrl = f.fallbackTarget
+		} else {
+			// TODO panic? - there is no way to recover from this point
+			log.Error("redis coordinator not initilized, no fallback available")
+			return f.retryAfterError()
+		}
+	}
+	if newSequencerUrl == f.currentTarget {
+		return nextUpdateIn()
+	}
+	var newForwarder *TxForwarder
+	for {
+		newForwarder = NewForwarder(newSequencerUrl, f.config)
+		err := newForwarder.Initialize(ctx)
+		if err == nil {
+			break
+		}
+		if f.shouldFallbackToStatic() && newSequencerUrl != f.fallbackTarget {
+			log.Error("failed to initialize forward agent, falling back to static url", "err", err, "fallback", f.fallbackTarget)
+			newSequencerUrl = f.fallbackTarget
+			nextUpdateIn = f.retryAfterError
+		} else {
+			log.Error("failed to initialize forward agent", "err", err)
+			return f.retryAfterError()
+		}
+	}
+	oldForwarder := f.getForwarder()
+	if oldForwarder != nil {
+		oldForwarder.Disable()
+	}
+	f.currentTarget = newSequencerUrl
+	f.setForwarder(newForwarder)
```

Comment:
Instead of disabling the old forwarder here we should do it in the setForwarder function.